spirv_reflect.c

```diff
@@ -3221,6 +3233,106 @@ static SpvReflectResult TraverseCallGraph(SpvReflectPrvParser* p_parser, SpvRefl
+    return not_found;
+  } else if (p_binding->user_type != SPV_REFLECT_USER_TYPE_BYTE_ADDRESS_BUFFER &&
+             p_binding->user_type != SPV_REFLECT_USER_TYPE_RW_BYTE_ADDRESS_BUFFER) {
+    return not_found;
+  }
+
+  uint32_t base_id = 0;
+  // expect first index of 2D access is zero
+  UNCHECKED_READU32(p_parser, p_node->word_offset + 4, base_id);
+  if (GetUint32Constant(p_parser, base_id) != 0) {
+    return not_found;
+  }
+  UNCHECKED_READU32(p_parser, p_node->word_offset + 5, base_id);
+  SpvReflectPrvNode* p_next_node = FindNode(p_parser, base_id);
+  if (IsNull(p_next_node)) {
+    return not_found;
+  }
+
+  // there is usually 2 (sometimes 3) instrucitons that make up the arithmetic logic to calculate the offset
+  SpvReflectPrvNode* arithmetic_node_stack[5];
+  uint32_t arithmetic_count = 0;
+
+  while (IsNotNull(p_next_node)) {
+    if (p_next_node->op == SpvOpLoad || p_next_node->op == SpvOpBitcast) {
+      break;  // arithmetic starts here
+    }
+    arithmetic_node_stack[arithmetic_count++] = p_next_node;
+    if (arithmetic_count > 5) {
+      return SPV_REFLECT_RESULT_ERROR_ALLOC_FAILED;
+    }
+
+    UNCHECKED_READU32(p_parser, p_next_node->word_offset + 3, base_id);
+    p_next_node = FindNode(p_parser, base_id);
+  }
+
+  uint32_t offset = 0;  // starting offset
+  const uint32_t count = arithmetic_count;
+  for (uint32_t i = 0; i < count; i++) {
+    p_next_node = arithmetic_node_stack[--arithmetic_count];
+    // All arithmetic ops takes 2 operands, assumption is the 2nd operand has the constant
+    UNCHECKED_READU32(p_parser, p_next_node->word_offset + 4, base_id);
+    uint32_t value = GetUint32Constant(p_parser, base_id);
+    if (value == INVALID_VALUE) {
+      return not_found;
+    }
+
+    switch (p_next_node->op) {
+      case SpvOpShiftRightLogical:
+        offset >>= value;
+        break;
+      case SpvOpIAdd:
+        offset += value;
+        break;
+      case SpvOpISub:
+        offset -= value;
+        break;
+      case SpvOpIMul:
+        offset *= value;
+        break;
+      case SpvOpUDiv:
+        offset /= value;
+        break;
+      case SpvOpSDiv:
+        // OpConstant might be signed, but value should never be negative
+        assert((int32_t)value > 0);
+        offset /= value;
+        break;
+      default:
+        return not_found;
+    }
+  }
+
+  // It is costly to get the size before, so trade off a larger memory footprint (only if SPIR-V has ByteAddressBuffer user type)
+  if (IsNull(p_binding->byte_address_buffer_offsets)) {
+    p_binding->byte_address_buffer_offsets = (uint32_t*)calloc(SPV_REFLECT_MAX_BYTE_ADDRESS_BUFFER, sizeof(uint32_t));
+    if (IsNull(p_binding->byte_address_buffer_offsets)) {
+      return SPV_REFLECT_RESULT_ERROR_ALLOC_FAILED;
+    }
+  } else if (p_binding->byte_address_buffer_offset_count >= SPV_REFLECT_MAX_BYTE_ADDRESS_BUFFER) {
+    return SPV_REFLECT_RESULT_ERROR_ALLOC_FAILED;
```

Comment:
I am hitting this case with one of my shaders causing a downstream crash and byte address buffer reflection info not to work.  This needs to not be a hard limit or at least must larger.

Reply:
the allocation is now dynamic and should work now no matter how large your ByteAddressBuffer is